carapace-server/src/main/java/org/carapaceproxy/server/RequestHandler.java

```diff
@@ -224,41 +223,64 @@ public void start() {
             case REDIRECT:
                 return;
             case PROXY: {
-                EndpointConnection connection;
-                try {
-//                    LOG.log(Level.SEVERE, "TIME"+TIME_TRACKER.incrementAndGet()+" start " + this + " thread " + Thread.currentThread().getName());
-                    connection = connectionToClient.connectionsManager.getConnection(new EndpointKey(action.host, action.port));
-                } catch (EndpointNotAvailableException err) {
-                    fireRequestFinished();
-                    LOG.log(Level.INFO, "{0} error on endpoint {1}: {2}", new Object[]{this, action, err});
-                    return;
+                connectionToEndpoint.set(connectionToClient.getConnectionToEndpoint()); // existing client2endpoint connection
+                if (connectionToEndpoint.get() == null) {
+                    try {
+                        connectionToEndpoint.set(connectionToClient.connectionsManager.getConnection(new EndpointKey(action.host, action.port)));
+                        connectionToClient.setConnectionToEndpoint(connectionToEndpoint.get());
+                        channelToClient.channel().closeFuture().addListener((Future<? super Void> future) -> {
+                            CarapaceLogger.debug("Closing channel to client id={0}.", connectionToClient.getId());
+                            EndpointConnection endpointConn = connectionToEndpoint.get();
```

Comment:
Endpoint connection binded to the request. Whether multiple connections have been opened by the client we gonna close them all.